PM_DEVICE_POWER_DOMAIN refactor

bjarki-andreasen opened the pull request:
This PR refactors PM_DEVICE_POWER_DOMAIN to be simpler and more scalable, moving power domain and device constraint logic from device drivers to PM_DEVICE_RUNTIME.

Depends on and cherry picks the following:
* https://github.com/zephyrproject-rtos/zephyr/pull/90938
* https://github.com/zephyrproject-rtos/zephyr/pull/90941
* https://github.com/zephyrproject-rtos/zephyr/pull/90946
* https://github.com/zephyrproject-rtos/zephyr/pull/90949

bjarki-andreasen-nordic committed:
drivers: power_domain: gpio: use pm_device_children_action_run

The power_domain_gpio driver was reimplementing the behavior of
pm_device_children_action_run(). Replace reimplementation with
pm_device_children_action_run().

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
pm: device: introduce helpers for turning on/off domains

Introduce helpers pm_device_domain_children_turn_on() and
pm_device_domain_children_turn_off() to be called from
PM_DEVICE_RUNTIME when after resuming and before suspending a
power domain.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
pm: device: runtime: implement pd and pm state logic

Move device power domain and pm policy device logic to
PM_DEVICE_RUNTIME. Specifically:

- When resuming a device, if successful, get pm policy power
  lock for device, and notify any device power domain children
  of the new state of the device by calling TURN_ON.
- When suspending a device, first notify power domain children
  that device will be suspended by calling TURN_OFF, then put
  pm policy device lock, then try to suspend the device. If
  failure, notify children that device will remain resumed,
  and get pm policy device lock again.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: power_domain: gpio: remove child actions:

PM_DEVICE_RUNTIME now implements power domain child actions. Remove
this duplicate logic from the device power domain driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: power_domain: intel_adsp: remove child actions:

PM_DEVICE_RUNTIME now implements power domain child actions. Remove
this duplicate logic from the device power domain driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
pm: policy: fix mocks of CONFIG_PM_POLICY_DEVICE_CONSTRAINTS

The APIs pm_policy_device_power_lock_get and
pm_policy_device_power_lock_put are not mocked properly as they
should be excluded based on CONFIG_PM_POLICY_DEVICE_CONSTRAINTS,
not CONFIG_PM.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: power_domain: add default power domain driver

Add default power domain driver which is useful for locking system
sleep states based on a group of devices.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: gpio: smartbond: fix PM_DEVICE_DEFINE usage

The gpio_smartbond driver incorrectly uses the static
PM_DEVICE_DEFINE() and PM_DEVICE_GET() macros when creating a driver
instance from a devicetree instance number.

Update to use PM_DEVICE_DT_INST_DEFINE() and
PM_DEVICE_DT_INST_GET() macros.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

decsny commented on dts/bindings/test/vnd,device-base.yaml:
I think this is what `vnd,reserved-compat` is specifically for already, so this new one shouldn't be needed

bjarki-andreasen replied:
nice, changed to use existing binding

bjarki-andreasen replied:
`vnd,reserved-node` node is specifically used to test devices with status reserved :) Added back the device-base one

decsny replied:
where did you see that?

bjarki-andreasen replied:
https://github.com/zephyrproject-rtos/zephyr/blob/e96bf44d934050b404e53aefdcb9756290b4c18e/tests/lib/devicetree/api/src/main.c#L377

decsny commented on include/zephyr/devicetree.h:
```diff
@@ -2893,6 +2893,91 @@
+	DT_CAT(node_id, _POWER_DOMAIN_CHILDREN_FOREACH_OKAY_SEP)(fn, sep)
+
+/**
+ * @brief Variant of DT_POWER_DOMAIN_CHILDREN_FOREACH_OKAY with VARGS
+ *
+ * @see DT_POWER_DOMAIN_CHILDREN_FOREACH_OKAY()
+ *
+ * @details The macro @p fn must take one parameter, which will be a node
+ * identifier, followed by variadic args. The macro is expanded once for
+ * every power domain child with status "okay".
+ *
+ * @param node_id Power domain's node identifier
+ * @param fn Macro to invoke
+ * @param ... Variadic args passed to @p fn
+ */
+#define DT_POWER_DOMAIN_CHILDREN_FOREACH_OKAY_VARGS(node_id, fn, ...) \
+	DT_CAT(node_id, _POWER_DOMAIN_CHILDREN_FOREACH_OKAY_VARGS)(fn, __VA_ARGS__)
```
other subsystems put the DT macros in include/zephyr/devicetree/ subfolder with their own header, can't this go that route as well?

bjarki-andreasen replied:
moved to its own header

decsny reviewed (dismissed):
The PR looks really good to me. The change request is specifically over some DT related nits. And I have one question about one of the other power domain drivers which is not being touched on this PR at the moment (power_domain_soc_state_change). Because that one in particular seems like the biggest refactor target of any of the existing in tree ones, given the goal of this PR is to remove code about calling turn on/off actions on domain children. That's basically all that driver's code seems to be for from what I can tell.

decsny commented:
And one more question , is are you targeting this PR to merge before 4.2 freeze point?

bjarki-andreasen commented:
@decsny I'm all for refactoring the soc_state_change and the board using it as part of this PR, it is indeed "backwards" compared to the power domain behavior I envision with this PR. I just did not want to "rock the boat" too much at once :)

bjarki-andreasen commented:
> And one more question , is are you targeting this PR to merge before 4.2 freeze point?

I'm working with an "as soon as possible" timeline to refactor pm device in general, so probably yes :D

bjarki-andreasen commented on scripts/dts/gen_defines.py:
```diff
@@ -736,6 +737,47 @@ def fmt_dep_list(dep_list):
+
+    out_comment("Node's power domain children:")
+
+    out_dt_define(f"{node.z_path_id}_POWER_DOMAIN_CHILDREN_LEN_OKAY", len(pd_child_z_path_ids))
+    out_dt_define(f"{node.z_path_id}_POWER_DOMAIN_CHILDREN_FOREACH_OKAY(fn)",
+                  " ".join(
+                  f"fn(DT_{z_path_id})"
+                  for z_path_id in pd_child_z_path_ids))
+    out_dt_define(f"{node.z_path_id}_POWER_DOMAIN_CHILDREN_FOREACH_OKAY_SEP(fn, sep)",
+                  " DT_DEBRACKET_INTERNAL sep ".join(
+                  f"fn(DT_{z_path_id})"
+                  for z_path_id in pd_child_z_path_ids))
+    out_dt_define(f"{node.z_path_id}_POWER_DOMAIN_CHILDREN_FOREACH_OKAY_VARGS(fn, ...)",
+                  " ".join(
+                  f"fn(DT_{z_path_id}, __VA_ARGS__)"
+                  for z_path_id in pd_child_z_path_ids))
+    out_dt_define(f"{node.z_path_id}_POWER_DOMAIN_CHILDREN_FOREACH_OKAY_SEP_VARGS(fn, sep, ...)",
+                  " DT_DEBRACKET_INTERNAL sep ".join(
+                  f"fn(DT_{z_path_id}, __VA_ARGS__)"
+                  for z_path_id in pd_child_z_path_ids))
+
```
added entry to macros.bnf

bjarki-andreasen-nordic committed:
samples: sensor: thermometer: add pm device runtime support

Add support for pm device runtime to sample. Simply resume the
temperature device before using it. If PM_DEVICE_RUNTIME is not
enabled, device is assumed resumed on boot.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: sensor: nxp: p3t1755: impl pm device runtime

Implement pm device runtime for p3t1755 sensor.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
@decsny I have refactored the rw612 to use "internal soc power domains" :) There are two power domains required, one limiting to PM2, one limiting to PM3. These have been added and devices have been updated to use them.

On top of this, the temperature sensor sample has been extended to implement PM_DEVICE_RUNTIME, which is required for device power domains and management to work.

Also, the temperature sensor driver used for the sample on the frdm_rw612 has been extended with support for PM_DEVICE_RUNTIME as well. I'm getting an frdm_rw612 board tomorrow to test it out.

Keep in mind that this only fixes the one sample in tree which enables PM and PM_DEVICE for the rw612, drivers still need to be updated to implement PM_DEVICE_RUNTIME and the frdm_rw612 board needs to be cleaned up, so it does not enable nodes which are not actually used :)

decsny commented:
> @decsny I have refactored the rw612 to use "internal soc power domains" :) There are two power domains required, one limiting to PM2, one limiting to PM3. These have been added and devices have been updated to use them.

I'm still a bit confused what's the idea behind this, because, again there are still domains corresponding to modes instead of actual relationships on the SOC? And still some of the nodes in that nxp_rw6xx_common.dtsi file are putting the zephyr,disabling-power states on the peripheral node, which I thought was to be avoided. 

Basically, AFAIK, for this chip, all digital peripherals are on the same domain, which will put them in retention (essentially on but not clocked and not functioning) in PM2, and off (lose state) in PM3. I'm not sure which of those corresponds to "disabled" in the DT prop terms, but my guess is both, so there probably should only be one domain instead of two. Any peripherals in the DT that previously only specifies PM3 as disabling is therefore probably a mistake.

There are a couple exceptions that are actually programmable as far as whether some domains affect them, but actually they are not described properly in DT or handled code right now anyways, and by default they behave the same as everything else as I described above. So NXP maintainers will deal with those later. But this programmability might seriously complicate the idea of domains in DT or app code, so before that I would like to make sure if we are clear on the concept of what a domain is. 

Correct me if you think differently, but right now I think it makes most sense based on what you've said for this PR, to have one domain for that chip (peripheral-domain ?)  and put everything on it, and disabling states would be suspend and standby.

> Keep in mind that this only fixes the one sample in tree which enables PM and PM_DEVICE for the rw612, drivers still need to be updated to implement PM_DEVICE_RUNTIME and the frdm_rw612 board needs to be cleaned up, so it does not enable nodes which are not actually used :)

Yes I know, that's perfectly fine. Right now we are doing a big push to write PM device enabling code in many of those drivers. There are some PRs open already actually. Although we're finding that it feels a bit unclear what the responsibilities of PM are at each layer of the system, but that's a topic for a different github thread than this PR, I guess.

bjarki-andreasen-nordic committed:
tests: pm: power_domain: use default power domain driver

Use the default power domain driver instead of creating a bespoke
one within the test suite.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
dts: arm: nxp: rw6xx_common.dtsi: use default power domain

Update board to use default power domain for the "peripheral"
domain, which when resumed prevents SoC from entering suspend and
standby states as these power down peripherals.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
samples: sensor: thermometer: frdm_rw612: enable PM_DEVICE_RUNTIME

Enable PM_DEVICE_RUNTIME and explicitlly enable dependencies
required for the power domains and device drivers power management
to function. For now, this requires manually enabling
PM_DEVICE_RUNTIME for every device used in the sample by adding the
devicetree prop zephyr,pm-device-runtime-auto to make the drivers
actually adhere to PM_DEVICE_RUNTIME...

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
> > @decsny I have refactored the rw612 to use "internal soc power domains" :) There are two power domains required, one limiting to PM2, one limiting to PM3. These have been added and devices have been updated to use them.
> 
> I'm still a bit confused what's the idea behind this, because, again there are still domains corresponding to modes instead of actual relationships on the SOC? And still some of the nodes in that nxp_rw6xx_common.dtsi file are putting the zephyr,disabling-power states on the peripheral node, which I thought was to be avoided.

I missed one, addressed :)

> Basically, AFAIK, for this chip, all digital peripherals are on the same domain, which will put them in retention (essentially on but not clocked and not functioning) in PM2, and off (lose state) in PM3. I'm not sure which of those corresponds to "disabled" in the DT prop terms, but my guess is both, so there probably should only be one domain instead of two. Any peripherals in the DT that previously only specifies PM3 as disabling is therefore probably a mistake.
> 
> There are a couple exceptions that are actually programmable as far as whether some domains affect them, but actually they are not described properly in DT or handled code right now anyways, and by default they behave the same as everything else as I described above. So NXP maintainers will deal with those later. But this programmability might seriously complicate the idea of domains in DT or app code, so before that I would like to make sure if we are clear on the concept of what a domain is.
> 
> Correct me if you think differently, but right now I think it makes most sense based on what you've said for this PR, to have one domain for that chip (peripheral-domain ?) and put everything on it, and disabling states would be suspend and standby.

Sounds right, moved all peripherals to a single domain, `peripheral_domain`, which prevents sleep and standby states. I just followed what was already in the dts file originally, and cross referenced it loosely with the datasheet.

> > Keep in mind that this only fixes the one sample in tree which enables PM and PM_DEVICE for the rw612, drivers still need to be updated to implement PM_DEVICE_RUNTIME and the frdm_rw612 board needs to be cleaned up, so it does not enable nodes which are not actually used :)
> 
> Yes I know, that's perfectly fine. Right now we are doing a big push to write PM device enabling code in many of those drivers. There are some PRs open already actually. Although we're finding that it feels a bit unclear what the responsibilities of PM are at each layer of the system, but that's a topic for a different github thread than this PR, I guess.

Yes please, this one https://github.com/zephyrproject-rtos/zephyr/issues/84049 preferably :)

decsny commented on dts/arm/nxp/nxp_rw6xx_common.dtsi:
```diff
@@ -300,8 +296,7 @@
 		dma-names = "rx", "tx";
-		zephyr,disabling-power-states = <&suspend &standby>;
-		power-domains = <&power_mode3_domain>;
+		power-domains = <&power_mode2_domain>;
```
looks like there are a few holdovers with a nonexistent domain

bjarki-andreasen replied:
ah, just did a replace all for `power_mode3_domain`, forgot the 2

bjarki-andreasen replied:
fixed

ceolin commented on drivers/power_domain/power_domain_intel_adsp.c:
```diff
@@ -63,20 +63,10 @@ static int pd_intel_adsp_pm_action(const struct device *dev, enum pm_device_acti
-		if (ret == 0) {
-			pm_device_children_action_run(dev, PM_DEVICE_ACTION_TURN_ON, NULL);
-		}
-
```
This can't be done until we have removed `PM_DEVICE_SYSTEM_MANAGED`. Otherwise if the target is built without device runtime children nodes will not get `TURN_ON` and `TURN_OFF` actions.

We will do this, but it has to be done after the `SYSTEM_MANAGED` is deprecated / removed.

ceolin commented on subsys/pm/device.c:
```diff
@@ -207,6 +207,58 @@ void pm_device_children_action_run(const struct device *dev,
+
+		LOG_ERR("%s could not be %s", child->name, "turned off");
+	}
+}
```
We should have only one function that receives the action to take. These two functions are doing pretty much the same thing with different actions. The main reason for it is to reduce ROM / flash usage.

bjarki-andreasen replied:
optimized to call single onoff function

bjarki-andreasen replied:
https://github.com/zephyrproject-rtos/zephyr/blob/e96bf44d934050b404e53aefdcb9756290b4c18e/tests/lib/devicetree/api/src/main.c#L377

JordanYates reviewed (changes requested):
Can the bugfixes and new feature support be pulled out to separate PR's please?
For a fundamental infrastructure change it just adds a bunch of noise.

Am I correct in saying that the core of this change is replacing the device ordinals with full device pointers?

bjarki-andreasen commented:
> Can the bugfixes and new feature support be pulled out to separate PR's please? For a fundamental infrastructure change it just adds a bunch of noise.

Sure, though they would still be part of this PR until merged individually.
 
> Am I correct in saying that the core of this change is replacing the device ordinals with full device pointers?

Its half of the core, the other half is: _"moving power domain and device constraint logic from device drivers to PM_DEVICE_RUNTIME" - title of PR_ :)

JordanYates commented:
> Its half of the core, the other half is: "moving power domain and device constraint logic from device drivers to PM_DEVICE_RUNTIME" - title of PR :)

Well the title is `PM_DEVICE_POWER_DOMAIN refactor`, and unless I missed it this PR doesn't change the logic around https://github.com/zephyrproject-rtos/zephyr/blob/4a31b8036d9c20349a0ee9ab83fa86147fdf983e/subsys/pm/device_runtime.c#L209-L212 Changes to the above are what I would have expected from anything that "moves" the power domain logic, so I wasn't entirely sure what was going on here. I'm still not clear what the description means in this context.

JordanYates commented:
Is that commit not applying the actions to all children of every device?
So for example, resuming a SPI controller would now run `ACTION_TURN_ON` for all SPI devices on that bus, even if they aren't powered?

JordanYates commented on include/zephyr/pm/device.h:
```diff
@@ -205,13 +211,57 @@ BUILD_ASSERT(offsetof(struct pm_device_isr, base) == 0);
+	static const struct device *Z_PM_DEVICE_DOMAIN_CHILDREN_NAME(dev_id)			\
+	[Z_PM_DEVICE_DOMAIN_DYNAMIC_NUM];
+
+#define Z_PM_DEVICE_DEFINE_DOMAIN_CHILDREN_DT(node_id, dev_id)					\
```
This was one of the reasons for using device ordinals in the original implementation (the other being the ordinals are half the size). Using integers allows the initial link stage to complete successfully, at which point our hacky scripts can run a second pass using the first linker output to determine which devices actually exist in the build.

bjarki-andreasen replied:
That breaks yes, just like any device in zephyr with DT_DEVICE_GET()

bjarki-andreasen replied:
Note that its not any child device, only child devices which point the parent with `power-domains = <>`. This power domain has to have a driver for the child device to work so the build would fail anyway if we ignored that no driver was applied to the power domain :)

bjarki-andreasen replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/89372/commits/4bd44cf6d229e857cf4ab11fea349d2d5094c1b7

JordanYates replied:
> That breaks yes, just like any device in zephyr with DT_DEVICE_GET()

So a regression in behaviour. The standard control for drivers is to enable the subsystem, whereas this adds a significant new requirement. Imagine a board with an internal power domain which the ADC hangs off. Suddenly either `CONFIG_ADC=n` breaks all builds, or `CONFIG_ADC=y` also requires a board specific overlay to actually work. The same logic applies to external power domains and other subsystems as well.

Not an acceptable state of affairs in my view.

> Using DEVICE_DEPS, a missing driver would be silently ignored at runtime.

Thats just describing the behavior of any piece of hardware without a compiled in driver, of course it is ignored at runtime.

> We spent additional time iterating every child, checking if that childs pm domain was the parent

I suspect the number of cases where a devicetree node is a child of a power domain node, but not because of a `power-domain = <>` node, is essentially 0. So there is basically no overhead, because in 99% of cases the two lists would just be the same.

bjarki-andreasen replied:
I see, I will revert back to using DEVICE_DEPS DEVICE_DEPS_DYNAMIC to track children.

My future vision is to move to a single linked list, and have children register themselves from `pm_device_driver_init()`, this removes the whole DYNAMIC_NUM deal, and makes it trivial to remove a device from a domain, like in the case of device_deinit() :) This of course requires all drivers to actually use `pm_device_driver_init()` so we can take a look at it again then :)

bjarki-andreasen commented:
> Is that commit not applying the actions to all children of every device? So for example, resuming a SPI controller would now run `ACTION_TURN_ON` for all SPI devices on that bus, even if they aren't powered?

no, only the child device pointed to by `power-domains = <>`.

bjarki-andreasen-nordic committed:
pm: device: pm_device_children_action_run: validate child domain

Validate that device dep children indeed belong to domain when
iterating dependent devices from pm_device_children_action_run().

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
### Update:

- Reverted changes to power domain child tracking, along with changes required to RW612 to deal with the enabled but unused nodes
- Enabled PM_DEVICE_RUNTIME for RW612 used in `samples/sensor/thermometer` using devicetree overlay

Currently testing on the RW612 to make sure it works as expected with the `samples/sensor/thermometer`, I will add a pm_device_runtime_stats feature similar to pm_stats to track devices states, checking how long they are resumed, which should help diagnose which devices are keeping an SoC from entering specific sleep states :)

decsny commented:
@bjarki-andreasen I don't know if this PR is the right place to bring this up, but I think we require a distinction between "turned off" and "disabled" for devices, and power domains seem like they might be the right place to handle this.

What I am understanding right now about the general spirit of the whole PM rework as in #89031 is that we want to decouple the state of the devices from the state of the system, ie suspend and resume decisions should not have anything to do with system mode transitions. But obviously it's recognized that devices are affected by the system modes in hardware nonetheless, and it appears that we are making the power domains the architectural glue between device and system. 

One of the main limitations that I am seeing with the architecture (current and proposed) for our chip, RW612, basically seems to mostly boil down to the fact that there is no distinction between devices that are powered off and losing state and devices which are disabled but not losing state due to the properties of the hardware domain they are on as it relates to the system modes.

I am seeing that we basically have these valid situations with your proposal (correct me if I included or excluded a situation by mistake, except for other system power levels, I am only interested in active, suspend to idle, and standby) for an arbitrary device on a domain (which may have other devices) within an SOC (system):

| System Mode | Domain | Device Status (software) | Device Status (hardware) |
|---|---|---|---|
| Active | On | Active | Active |
| Active | On | Suspended | Up to driver |
| Active | Off | Off | Lost state |
| Suspend | On | Active | Active |
| Suspend | On | Suspended | Up to driver |
| Suspend | Off | Off | Lost state |
| Standby | On | Suspended | Up to driver |
| Standby | Off | Off | Lost state |

I am going by the description of the power states right now which says that suspend to idle *may* put peripherals in low power state (I am assuming this means domains may or may not be turned off then), and standby does put all peripherals into low power states (except maybe some exceptions, I recognize that it gets a bit hairy per platform).

The issue here for our chip (RW612) is that there is two levels of effect by two different low power modes, the first will disable the peripherals to conserve power but not lose state, and the second will depower completely the peripherals and they will lose state.

The issue here then is that since "suspend" is now meant to be almost orthogonal to system transitions, we can't use it to mean "disabled", since this has nothing to do with software control and actually the device driver's handling of "suspend" might do different things on a device level than what happens at a system level where the entire domain is just disabled (you can think of it maybe conceptually as all the clocks are gated, which is part of what actually happens). From my understanding, we are trying to get away from system power transitions causing devices to suspend. Probably the driver when it gets a suspend action should disable its own clock. But I think what we need to know is if some external event caused the device to be disabled due to system wide state even if it wasnt suspended at the device/domain level.

So I am not sure exactly how to handle this, but one idea is that another sort of state or action is added to the PM device or domain state machine, so then the valid possible situations would expand at least to this (not necessarily all possible on every platform, depends on domain glue):

| System Mode | Domain | Device Status (software) | Device Status (hardware) |
|---|---|---|---|
| Active | On | Active | Active |
| Active | On | Suspended | Up to driver |
| Active | Low | Disabled | Inactive but retain state |
| Active | Off | Off | Lost state |
| Suspend | On | Active | Active |
| Suspend | On | Suspended | Up to driver |
| Suspend | Low | Disabled | Inactive but retain state |
| Suspend | Off | Off | Lost state |
| Standby | On | Suspended | Up to driver |
| Suspend | Low | Disabled | Inactive but retain state |
| Standby | Off | Off | Lost state |

bjarki-andreasen commented:
@decsny

The states are actually quite a bit simpler :)

X here means "don't care"

| System mode | domain | Device status (software) | Device status (hardware) |
|---|---|---|---|
| X | On | Active | Active |
| X | On | Suspended | Active |
| X | Off | Off | X |

All the power domain does is prevent a system from transitioning to a state which will disable the device hardware. When and if the system actually enters said state is not actually relevant to the device :) 

When the PM_DEVICE_RUNTIME calls suspend on a device, it means two things:
1. User no longer needs to use device, so go to the lowest possible power mode.
2. If device is on power domain, the power domain may now be suspended at any time, at which point the device hardware **may** loose power, so do act as if it already has.

Now, if the device is on a power domain, and all devices on the power domain are suspended, the power domain will be suspended. Once it is, it will release the system pm state lock if it has one, thus, and do note, no longer **preventing** the system from entering sleep modes. Child devices will get the turn_off action immediately when the power domain is suspended, not when and if the system actually transitions to a sleep state.

The system can have states locked for any number of reasons, like to meet some minimum latency requirement, every single device and power domain may be suspended, but if something has requested a pm policy latency of 0, the system will not enter a sleep state :)

With this said, a device may still opt to monitor for and react to pm state transitions if necessary, that is totally fine, but as the table above suggests, PM_DEVICE_RUNTIME does not care.

decsny commented:
> All the power domain does is prevent a system from transitioning to a state which will disable the device hardware. When and if the system actually enters said state is not actually relevant to the device (from the perspective of PM_DEVICE_RUNTIME) :)

Yes, that was my understanding. But what I'm saying is that for us there basically is what seems like more than 2 binary states (on, off) for our domain, there is also a situation where the domain is on but everything that is part of it is disabled but not off, which has real implications for the system, and these domain states in the hardware are affected by the system mode. Right now we have no way to differentiate at the domain level between a system state that will cause the domain to turn completely off vs cause the domain to be in this low power state that disables things. 

> When the PM_DEVICE_RUNTIME calls suspend on a device, it means two things:
> 
>     1. User no longer needs to use device, so go to the lowest possible power mode.
> 
>     2. If device is on power domain, the power domain may now be suspended at any time, at which point the device hardware **may** loose power, so do act as if it already has.

This was also my understanding but I do wonder what implication this has for driver development. Like how is the driver supposed to act or what should it do if it loses power, it's not clear. Right now for example on the uart driver we have, we save off which interrupts are enabled so that upon a turn on action, we re-enable them. Or is that wrong and app should reconfigure? Or should the driver do more than this? Because right now it just seems to be a wild west of band-aids of different varieties for different system designs and use cases. It's unclear to me but maybe outside the scope of this PR, although we are talking about the domain-related actions specifically (on and off).
 
> Now, if the device is on a power domain, and all devices on the power domain are suspended, the power domain will be suspended. Once it is, it will release the system pm state lock if it has one, thus, and do note, no longer **preventing** the system from entering sleep modes. Child devices will get the turn_off action immediately when the power domain is suspended, not when and if the system actually transitions to a sleep state.

Yes, I understood that as well, that is the best part of this PR. It makes the domains and device PM a lot more solid and clear. This is what I have been explaining that this PR is doing to other people.

> The system can have states locked for any number of reasons, like to meet some minimum latency requirement. Every single device and power domain may be suspended, but if something has requested a pm policy latency of 0, the system will not enter a sleep state, and all device hardware remains Active :)

I think now I am confused what you mean by active here, because I thought that if a power domain is "suspended" (as you call it, although I am confused by the terminology because I thought that domains are only either on or off), that all the devices on that domain are also off, or at least suspended, but I can see in the PM device state machine that off and suspended are different states, and I thought the off action is the one that a domain would cause... 

> With this said, a device may still opt to monitor for and react to pm state transitions if necessary, that is totally fine, but as the table above suggests, PM_DEVICE_RUNTIME does not care.

And I think you are referring to something like what I did on #89760, and I get that the monitoring of the system state transitions is fine, but, what I am trying to say, is that the constraints are not distinct enough. Right now the PM constraints come from the Devicetree property `zephyr,disabling-states`, but there is no way to tell apart in our case these states that will only disable the device (or domain) vs depower it. So what I am basically wondering is if this DT property should be split into two more specific ones, which are like `disabling-states` and `depowering-states`, and have those two separate types of PM constraints, which seem to already be based around system modes, and not related to runtime device PM.

JordanYates commented:
>> When the PM_DEVICE_RUNTIME calls suspend on a device, it means two things:

> 1. User no longer needs to use device, so go to the lowest possible power mode.
> 2. If device is on power domain, the power domain may now be suspended at any time, at which point the device hardware **may** loose power, so do act as if it already has.

This is pretty explicitly **NOT** the correct behaviour. `SUSPEND` requires that context is preserved, `OFF` is used when context is lost. Requiring all drivers to be updated such that `SUSPEND` **could** lose context is both a large piece of work and a bad idea IMO.

If a device is in `SUSPEND` and context is about to be lost, it is expected that `ACTION_TURN_OFF` is called beforehand.

decsny commented:
> > > When the PM_DEVICE_RUNTIME calls suspend on a device, it means two things:
> 
> > 1. User no longer needs to use device, so go to the lowest possible power mode.
> > 2. If device is on power domain, the power domain may now be suspended at any time, at which point the device hardware **may** loose power, so do act as if it already has.
> 
> This is pretty explicitly **NOT** the correct behaviour. `SUSPEND` requires that context is preserved, `OFF` is used when context is lost. Requiring all drivers to be updated such that `SUSPEND` **could** lose context is both a large piece of work and a bad idea IMO.
> 
> If a device is in `SUSPEND` and context is about to be lost, it is expected that `ACTION_TURN_OFF` is called beforehand.

This is why I was asking on my last comment what did he mean by suspend, I think the term is being overloaded. As far as I know this PR is not changing the meaning of suspend state for devices but changing the behavior of the domains so that if all of the children of the domain are suspended then it will also turn them all off.

decsny commented:
@JordanYates is your change request still relevant? This PR makes a lot of sense to me in it's current state, I would actually like to support it's merge sooner than later, so do we need an arbitration? 

@bjarki-andreasen I'm not sure what are the twister failures here but is this PR finished otherwise?

JordanYates commented on subsys/pm/device_runtime.c:
```diff
@@ -24,6 +25,31 @@ LOG_MODULE_DECLARE(pm_device, CONFIG_PM_DEVICE_LOG_LEVEL);
+static int invoke_suspend(const struct device *dev)
+{
+	pm_policy_device_power_lock_put(dev);
+	pm_device_domain_children_turn_off(dev);
```
AFAICT this pattern is still adding a chunk of overhead to all PM calls.

Every suspend/resume cycle is iterating over every child node regardless of whether the device has any capability of acting as a power domain. This is particularly problematic for `PM_DEVICE_FLAG_ISR_SAFE`, which is running all of these operations in an ISR context.

Should this not be using `PM_DEVICE_FLAG_PD` in order to determine whether to run these functions in the first place?
This would also allow dropping the `pm: device: pm_device_children_action_run: validate child domain` commit entirely, as I can't see any way in which a device could be a child of a power domain without being on that power domain.

JordanYates reviewed (changes requested):
This PR is still doing a bunch of things that are not coupled to each other:
 * Fixing a GPIO driver
 * Updating the implementation of existing power domains
 * Adding a new type of power domain
 * Multiple commits related to one particular sensor sample
 * Implementing PM on a random sensor
 * Changing power domain setups on one SoC

Not all of these need to be their own PR, but the core of this PR (based on the title) is changing where power domain logic runs. That can be done without any of the extra commits that have crept in here.

I don't think its that hard to split this up to be more manageable (and faster for the trivial aspects)

JordanYates commented on subsys/pm/device_runtime.c:
```diff
@@ -24,6 +25,31 @@ LOG_MODULE_DECLARE(pm_device, CONFIG_PM_DEVICE_LOG_LEVEL);
+	if (dev->pm_base->action_cb(dev, PM_DEVICE_ACTION_SUSPEND) < 0) {
+		pm_device_domain_children_turn_on(dev);
+		pm_policy_device_power_lock_get(dev);
+		return -EAGAIN;
```
This also overrides return values from the drivers, which was not previously the case

bjarki-andreasen replied:
return value now returned

JordanYates commented:
> which means devices relying on previous power domain logic need to be updated

I'm not complaining about the commits that update power domain drivers, they are obviously fine.
Its the assortment of other changes, some of which are the cause of CI failures.
Other parts (e.g. the PM policy changes) are not obvious that they are even changed in this PR from the title or description, and people that care about PM policy may have simply skipped the PR as not relevant to them.

bjarki-andreasen commented:
> > which means devices relying on previous power domain logic need to be updated
> 
> I'm not complaining about the commits that update power domain drivers, they are obviously fine. Its the assortment of other changes, some of which are the cause of CI failures. Other parts (e.g. the PM policy changes) are not obvious that they are even changed in this PR from the title or description, and people that care about PM policy may have simply skipped the PR as not relevant to them.

I get your point, I will create new PRs for those issues and cherry-pick them here :)

bjarki-andreasen-nordic committed:
pm: device_runtime: fix unbalanced domain get/put

pm_device_runtime has inconsistent behavior regarding
getting and putting a device's domain. This commit
aligns it to get the domain only once, once device
is resumed, and put only once, once device is
suspended.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
Please review the following individual PRs this PR depends on / cherry picks

* https://github.com/zephyrproject-rtos/zephyr/pull/90938
* https://github.com/zephyrproject-rtos/zephyr/pull/90941
* https://github.com/zephyrproject-rtos/zephyr/pull/90946
* https://github.com/zephyrproject-rtos/zephyr/pull/90949

bjarki-andreasen-nordic committed:
tests: counter: basic_api: Don't enable device PM

PM_DEVICE_SYSTEM_MANAGED nor PM_DEVICE_RUNTIME is not supported by
test suite, so disable any config which results in either being
enabled.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
samples: bluetooth: central_ht: Don't enable device PM

PM_DEVICE_SYSTEM_MANAGED nor PM_DEVICE_RUNTIME is not supported by
test suite, so disable any config which results in either being
enabled.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
currently I'm working towards standardizing on PM DEVICE RUNTIME with PRs like #93720 since power domains depend on PM DEVICE RUNTIME, so this PR is a bit less prioritized

decsny commented:
@bjarki-andreasen can you help me understand what's the status of this now? It seems like this PR is required to have a coherent PM device runtime management. We are kind of blocked waiting for this as we are wanting to enable multiple SOC right now in the paradigm you're trying to lead everybody towards.

It seems like some of the commits on this PR are actually merged from other PRs? Can we do a rebase and see what's actually left in this PR?
